cpp/src/copying/gather.cu

```diff
@@ -0,0 +1,106 @@
+#include "gather.cuh"
+#include <cudf/types.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <cudf/table/table_view.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/copying.hpp>
+#include <utilities/error_utils.hpp>
+
+#include <rmm/thrust_rmm_allocator.h>
+#include <thrust/count.h>
+
+#include <memory>
+
+
+namespace cudf {
+namespace experimental {
+namespace detail {
+
+
+struct dispatch_map_type {
+  template <typename map_type, std::enable_if_t<std::is_integral<map_type>::value>* = nullptr>
+  std::unique_ptr<table> operator()(table_view const& source_table,
+				    column_view const& gather_map,
+				    size_type num_destination_rows, bool check_bounds,
+				    bool ignore_out_of_bounds,
+				    bool allow_negative_indices = false)
+  {
+    std::unique_ptr<table> destination_table;
+    map_type const * typed_gather_map = gather_map.data<map_type>();
+
+    if (check_bounds) {
+      cudf::size_type begin = (allow_negative_indices) ? -source_table.num_rows() : 0;
+      CUDF_EXPECTS(
+	  num_destination_rows == thrust::count_if(
+	      rmm::exec_policy()->on(0),
+	      typed_gather_map,
+	      typed_gather_map + num_destination_rows,
+	      bounds_checker<map_type>{begin, source_table.num_rows()}),
+	  "Index out of bounds.");
+    }
+
+    if (allow_negative_indices) {
+      destination_table =
+	gather(source_table,
+	       thrust::make_transform_iterator(
+					       typed_gather_map,
+					       index_converter<map_type,index_conversion::NEGATIVE_TO_POSITIVE>{source_table.num_rows()}),
```

Comment:
Likewise, `index_converter` could be replace by a simple lambda. 
```suggestion
					       [n_rows](auto in){ return ((in % n_rows) + n_rows) % n_rows;}
```

Reply:
I'm running into some strange errors when trying to use a lambda instead (even after adding `--expt-extended-lambda` as suggested [here](https://stackoverflow.com/questions/30438538/lambda-expressions-with-cuda)).

Are device lambdas fully supported with Thrust?

Reply:
It needs to be marked a device lambda:
```
[n_rows] __device__ (auto in){ return ((in % n_rows) + n_rows) % n_rows;}
```

Reply:
`--expt-extended-lambda` should already be in the cmake CUDA settings. Is it not?

Reply:
It is - sorry. I just realized I duplicated it